src/kernels/jupyter/finder/remoteKernelFinder.ts

```diff
@@ -67,102 +92,137 @@ export class RemoteKernelFinder implements IRemoteKernelFinder, IExtensionSingle
         @inject(IJupyterRemoteCachedKernelValidator)
         private readonly cachedRemoteKernelValidator: IJupyterRemoteCachedKernelValidator,
         @inject(IKernelFinder) kernelFinder: KernelFinder,
+        @inject(IDisposableRegistry) private readonly disposables: IDisposableRegistry,
+        @inject(IKernelProvider) private readonly kernelProvider: IKernelProvider,
+        @inject(IExtensions) private readonly extensions: IExtensions,
         @inject(IsWebExtension) private isWebExtension: boolean
     ) {
+        this._initializedPromise = new Promise<void>((resolve) => {
+            this._initializeResolve = resolve;
+        });
+
         kernelFinder.registerKernelFinder(this);
     }
 
     async activate(): Promise<void> {
-        noop();
-    }
+        // warm up the cache
+        this.loadCache().then(noop, noop);
 
-    async listContributedKernels(
-        resource: Resource,
-        cancelToken: CancellationToken | undefined,
-        useCache: 'ignoreCache' | 'useCache'
-    ): Promise<KernelConnectionMetadata[]> {
-        const kernels: KernelConnectionMetadata[] = await this.listKernelsImpl(resource, cancelToken, useCache).catch(
-            (ex) => {
-                // Sometimes we can get errors from the socket level or jupyter, with the message 'Canceled', lets ignore those
-                if (!isCancellationError(ex, true)) {
-                    traceError('Failed to get remote kernels', ex);
-                }
-                return [];
+        this.disposables.push(
+            this.serverUriStorage.onDidChangeUri(() => {
+                this.updateCache().then(noop, noop);
+            })
+        );
+
+        // If we create a new kernel, we need to refresh if the kernel is remote (because
+        // we have live sessions possible)
+        // Note, this is a perf optimization for right now. We should not need
+        // to check for remote if the future when we support live sessions on local
+        this.kernelProvider.onDidStartKernel((k) => {
+            if (isRemoteConnection(k.kernelConnectionMetadata)) {
+                // update remote kernels
+                this.updateCache().then(noop, noop);
             }
+        });
+
+        // For kernel dispose we need to wait a bit, otherwise the list comes back the
+        // same
+        this.kernelProvider.onDidDisposeKernel(
+            (k) => {
+                if (k && isRemoteConnection(k.kernelConnectionMetadata)) {
+                    setTimeout(() => {
```

Comment:
Can we capture the return value so that the `timer` can be disposed in the `dipose` method.
This ensures we don't have any timers hanging around that could cause the node process from hanging instead of shutting down gracefully.
E.g. if i were to create this class and dispoose this immediately and end the tests, it would actually hang due to pending timers.

We've had a lot of issues in the past due to such dangling timers.